### PR TITLE
update to Changed-module build

### PR DESCRIPTION
**Pull request by Nickcandy**
## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by Nickcandy** [comment]
@isra-fel  @wyunchi-ms 
Hi, please help me review the above changes about the module build

Here is the test bump version: https://github.com/Azure/azure-powershell/pull/23332/files
Here is the test daily build: https://dev.azure.com/azclitools/internal/_build/results?buildId=105648&view=results
Here is the test sign: https://dev.azure.com/azclitools/release/_build/results?buildId=105643&view=results

**Review by VeryEarly — COMMENTED** [review]
is the empty changelog.md intended?

**Review by isra-fel — CHANGES_REQUESTED** [review]

**Comment by Nickcandy** [comment]
@isra-fel  @wyunchi-ms 
Hi, please help me review the above changes about the module build

Here is the test bump version: https://github.com/Azure/azure-powershell/pull/23332/files
Here is the test daily build: https://dev.azure.com/azclitools/internal/_build/results?buildId=109250&view=results
Here is the test sign: https://dev.azure.com/azclitools/release/_build/results?buildId=109297&view=results

File: tools/BuildScripts/CheckChangelogs.ps1 [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
+param(
+    [Parameter(Mandatory=$false)]
+    [string] $rootPath = ".",
+    [Parameter(Mandatory=$True)]
+    [string] $outputFile = "artifacts/ChangedModule.txt",
```
Comment (wyunchi-ms) [inline tools/BuildScripts/CheckChangelogs.ps1:18]: change default value to `artifacts/ModifiedModule.txt`

File: tools/BuildScripts/BuildModules.ps1 [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    if ($Configuration -ne 'Release' -and $TestsToRun -eq 'NonCore') {
+        $csprojFiles += Include-CsprojFiles -Path "$RepoRoot/src" -Exclude $CoreTests -Filter "*.Test.csproj"
+    }
+
+    if ($Configuration -ne 'Release' -and $TestsToRun -eq 'Core') {
+        $csprojFiles += Include-CsprojFiles -Path "$RepoRoot/src" -Include $CoreTests
+    }
+
+    if ($env:OS -eq "Windows_NT") {
+        $csprojFiles += Include-CsprojFiles -Path "$RepoRoot/src" -Filter "Authenticators.csproj"
```
Comment (wyunchi-ms) [inline tools/BuildScripts/BuildModules.ps1:47]: These code can be reused with line 57 - line 70

File: .azure-pipelines/code-sign.yml [hunk 3]
```diff
@@ -3,6 +3,14 @@ parameters:
   displayName: Branch for OOB pipeline
   type: string
   default: 'none'
+- name: ChangedModule
+  displayName: Build Module List(Split by ':')
```
[the remark this answers is not in the document]
Comment (Nickcandy) [reply .azure-pipelines/code-sign.yml:7]: Yes I know that. But It seems like `;` or `,` cannot use in dotnet cli after I test.

File: build.proj [hunk 4]
```diff
@@ -149,33 +153,18 @@
...
+    <Exec Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -BuildCsprojList '@(BuildCsprojList)' -TestCsprojList '@(TestCsprojList)'&quot;" Condition="$(PullRequestNumber) != ''"/>
+    <Exec Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -ModifiedModuleBuild $(ModifiedModuleBuild) &quot;"  Condition="$(ModifiedModuleBuild) != 'false'"/>
+    <Exec Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -TargetModule $(TargetModule)&quot;" Condition="$(TargetModule) != ''"/>
+
```
Comment (wyunchi-ms) [inline build.proj:167]:
```suggestion
    <Exec Condition="$(PullRequestNumber) == '' AND $(TargetModule) == '' AND $(ModifiedModuleBuild) == 'false'"
      Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) &quot;"
    />
    <Exec Condition="$(PullRequestNumber) != ''"
      Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -BuildCsprojList '@(BuildCsprojList)' -TestCsprojList '@(TestCsprojList)'&quot;"
    />
    <Exec Condition="$(ModifiedModuleBuild) != 'false'"
      Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -ModifiedModuleBuild $(ModifiedModuleBuild) &quot;"
    />
    <Exec Condition="$(TargetModule) != ''"
      Command="$(PowerShellCoreCommandPrefix) &quot;.\tools\BuildScripts\BuildModules.ps1 -RepoRoot $(RepoRoot) -Configuration $(Configuration) -TestsToRun $(TestsToRun) -CoreTests $(CoreTests) -RepoArtifacts $(RepoArtifacts) -TargetModule $(TargetModule)&quot;"
    />
```

File: tools/BuildScripts/BuildModules.ps1 [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            $_
+        }
+    }
+    
+    $allItems = Get-ChildItem -Path $Path -Filter $Filter -Recurse
+    $includedFiltered = $allItems | Where-Object { $includeItems -contains $_.FullName }
+    $finalFiltered = $includedFiltered | Where-Object { $excludeItems -notcontains $_.FullName}
+    $finalFiltered
+}
+
```
Comment (wyunchi-ms) [inline tools/BuildScripts/BuildModules.ps1:49]:
```suggestion
function Add-TestProjects {
    param (
        [string[]] $csprojFiles,
        [string]$TestsToRun,
        [string]$CoreTests,
        [string]$modulePath
    )
        $csprojFiles += Include-CsprojFiles -Path "$modulePath" -Exclude "*.Test.csproj;Authenticators.csproj" -Filter "*.csproj"
        if ($Configuration -ne 'Release') {
            if ($TestsToRun -eq 'All') {
                $csprojFiles += Include-CsprojFiles -Path $modulePath -Filter "*.Test.csproj"
            } elseif ($TestsToRun -eq 'NonCore') {
                $csprojFiles += Include-CsprojFiles -Path $modulePath -Exclude $CoreTests -Filter "*.Test.csproj"
            } elseif ($TestsToRun -eq 'Core') {
                $csprojFiles += Include-CsprojFiles -Path $modulePath -Include $CoreTests
            }
        }

        if ($env:OS -eq "Windows_NT") {
            $csprojFiles += Include-CsprojFiles -Path "$modulePath" -Filter "Authenticators.csproj"
        }
}
```

**Review by wyunchi-ms — APPROVED** [review]

**Pull request merged**